src/data/padics/ring_homs.lean

```diff
@@ -0,0 +1,689 @@
+  have : ↑p ∣ r.num ∧ (p : ℤ) ∣ r.denom,
+  { simp only [← norm_int_lt_one_iff_dvd, ← padic_norm_e_of_padic_int],
+    norm_cast, exact ⟨key, norm_denom_lt⟩ },
+  apply hp_prime.not_dvd_one,
+  rwa [← r.cop.gcd_eq_one, nat.dvd_gcd_iff, ← int.coe_nat_dvd_left, ← int.coe_nat_dvd],
+end
+
+lemma norm_sub_mod_part_aux (r : ℚ) (h : ∥(r : ℚ_[p])∥ ≤ 1) :
+  ↑p ∣ r.num - r.num * r.denom.gcd_a p % p * ↑(r.denom) :=
+begin
+  rw ← zmod.int_coe_zmod_eq_zero_iff_dvd,
+  simp only [int.cast_coe_nat, zmod.cast_mod_nat p, int.cast_mul, int.cast_sub],
+  have := congr_arg (coe : ℤ → zmod p) (gcd_eq_gcd_ab r.denom p),
+  simp only [int.cast_coe_nat, add_zero, int.cast_add, zmod.cast_self, int.cast_mul, zero_mul] at this,
+  push_cast,
+  rw [mul_right_comm, mul_assoc, ←this],
+  suffices rdcp : r.denom.coprime p,
+  { rw rdcp.gcd_eq_one, simp only [mul_one, cast_one, sub_self], },
+  apply coprime.symm,
+  apply (coprime_or_dvd_of_prime ‹_› _).resolve_right,
+  rw [← int.coe_nat_dvd, ← norm_int_lt_one_iff_dvd, not_lt],
+  apply ge_of_eq,
+  rw ← is_unit_iff,
+  exact is_unit_denom r h,
+end
+
+lemma norm_sub_mod_part (h : ∥(r : ℚ_[p])∥ ≤ 1) : ∥(⟨r,h⟩ - mod_part p r : ℤ_[p])∥ < 1 :=
+begin
+  let n := mod_part p r,
+  by_cases aux : (⟨r,h⟩ - n : ℤ_[p]) = 0,
+  { rw [aux, norm_zero], exact zero_lt_one, },
+  rw [norm_lt_one_iff_dvd, ← (is_unit_denom r h).dvd_mul_right],
+  suffices : ↑p ∣ r.num - n * r.denom,
+  { convert (int.cast_ring_hom ℤ_[p]).map_dvd this,
+    simp only [sub_mul, int.cast_coe_nat, ring_hom.eq_int_cast, int.cast_mul,
+      sub_left_inj, int.cast_sub],
+    apply subtype.coe_injective,
+    simp only [coe_mul, subtype.coe_mk, coe_coe],
+    rw_mod_cast @rat.mul_denom_eq_num r, refl },
+  dsimp [n, mod_part],
+  apply norm_sub_mod_part_aux r h,
+end
+
+lemma exists_mem_range_of_norm_rat_le_one (h : ∥(r : ℚ_[p])∥ ≤ 1) :
+  ∃ n : ℤ, 0 ≤ n ∧ n < p ∧ ∥(⟨r,h⟩ - n : ℤ_[p])∥ < 1 :=
+⟨mod_part p r, mod_part_nonneg _, mod_part_lt_p _, norm_sub_mod_part _ h⟩
+
+end
+
+section
+
+lemma zmod_congr_of_sub_mem_span_aux (n : ℕ) (x : ℤ_[p]) (a b : ℤ)
+  (ha : x - a ∈ (ideal.span {p ^ n} : ideal ℤ_[p]))
+  (hb : x - b ∈ (ideal.span {p ^ n} : ideal ℤ_[p])) :
+  (a : zmod (p ^ n)) = b :=
+begin
+  rw [ideal.mem_span_singleton] at ha hb,
+  rw [← sub_eq_zero, ← int.cast_sub,
+      zmod.int_coe_zmod_eq_zero_iff_dvd, int.coe_nat_pow],
+  rw [← dvd_neg, neg_sub] at ha,
+  have := dvd_add ha hb,
+  rwa [sub_eq_add_neg, sub_eq_add_neg, add_assoc, neg_add_cancel_left,
+      ← sub_eq_add_neg, ← int.cast_sub, pow_p_dvd_int_iff] at this,
+end
+
+lemma zmod_congr_of_sub_mem_span (n : ℕ) (x : ℤ_[p]) (a b : ℕ)
+  (ha : x - a ∈ (ideal.span {p ^ n} : ideal ℤ_[p]))
+  (hb : x - b ∈ (ideal.span {p ^ n} : ideal ℤ_[p])) :
+  (a : zmod (p ^ n)) = b :=
+zmod_congr_of_sub_mem_span_aux n x a b ha hb
+
+lemma zmod_congr_of_sub_mem_max_ideal (x : ℤ_[p]) (m n : ℕ)
+  (hm : x - m ∈ maximal_ideal ℤ_[p]) (hn : x - n ∈ maximal_ideal ℤ_[p]) :
+  (m : zmod p) = n :=
+begin
```

Comment:
It's a bit annoying that this can't be instantly deduced from the lemma before.

Reply:
I'll look again, but I think the reason is that `p^1` isn't defeq to `p`.

Reply:
Ah, yeah. This wouldn't be so bad if it were only `span {↑p^1} = span {↑p}` getting in the way, but there's also a type equality `zmod (p^1) = zmod p`. I also wish this were less of a contortion, but I don't see a slick way to improve it.